File: integration/BattleCityDemo/Assets/Scripts/behaviac/Agent/World.cs
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using UnityEngine;

namespace behaviac
{
    public class World : Agent
    {
        protected World()
        {
        }

        protected new void OnDestroy()
        {
            Context c = Context.GetContext(this.m_contextId);
            World pW = c.GetWorld(false);

            if (pW == this)
            {
                c.SetWorld(null);
            }

            base.OnDestroy();
        }

        protected new void Init()
        {
            Context c = Context.GetContext(this.m_contextId);
            c.SetWorld(this);

            base.Init();
            m_bTickAgents = true;
        }

        public static World GetInstance(int contextId)
        {
            Context c = Context.GetContext(contextId);
            World pWorld = c.GetWorld(true);

            Debug.Check(pWorld != null);

            return pWorld;
        }


        public struct HeapItem_t : IComparable<HeapItem_t>
        {
            public int priority;
            public Dictionary<int, Agent> agents;

            public int CompareTo(HeapItem_t other)
            {
                if (this.priority < other.priority)
                {
                    return -1;
                }
                else if (this.priority > other.priority)
                {
                    return 1;
                }

                return 0;
            }
        }

        List<HeapItem_t> m_agents;
        public List<HeapItem_t> Agents
        {
            get
            {
                if (m_agents == null)
                {
                    m_agents = new List<HeapItem_t>();
                }

                return m_agents;
            }
            set
            {
                m_agents = value;
            }
        }

        bool m_bTickAgents;

        public void AddAgent(Agent pAgent)
        {
            //ASSERT_MAIN_THREAD();
            int agentId = pAgent.GetId();
            int priority = pAgent.GetPriority();
            int itemIndex = this.Agents.FindIndex(delegate(HeapItem_t h) { return h.priority == priority; });

            if (itemIndex == -1)
            {
                HeapItem_t pa = new HeapItem_t();
                pa.agents = new Dictionary<int, Agent>();
                pa.priority = priority;
                pa.agents[agentId] = pAgent;
                this.Agents.Add(pa);
            }
            else
            {
                this.Agents[itemIndex].agents[agentId] = pAgent;
            }
        }

        public void RemoveAgent(Agent pAgent)
        {
            //ASSERT_MAIN_THREAD();

            int agentId = pAgent.GetId();
            int priority = pAgent.GetPriority();
            int itemIndex = this.Agents.FindIndex(delegate(HeapItem_t h) { return h.priority == priority; });

            if (itemIndex != -1)
            {
                if (this.Agents[itemIndex].agents.ContainsKey(agentId))
                {
                    this.Agents[itemIndex].agents.Remove(agentId);
                }
            }
        }

        public void RemoveAllAgents()
        {
            //ASSERT_MAIN_THREAD();

            this.Agents.Clear();
        }

        public override EBTStatus btexec()
        {
            //behaviac.Profiler.GetInstance().BeginFrame();
			Workspace.LogFrames ();
            Workspace.HandleRequests();

            if (Workspace.GetAutoHotReload())
            {
                Workspace.HotReload();
            }

            /*EBTStatus s = */base.btexec();

            if (this.m_bTickAgents)
            {
                this.btexec_agents();
            }

            //behaviac.Profiler.GetInstance().EndFrame();

            //TODO:return invalid or s?
            return EBTStatus.BT_RUNNING;
        }

        public void ToggleTickAgents(bool bTickAgents)
        {
            this.m_bTickAgents = bTickAgents;
        }

        public bool IsTickAgents()
        {
            return this.m_bTickAgents;
        }

        void btexec_agents()
        {
            this.Agents.Sort();

			for (int i = 0; i < this.Agents.Count; ++i)
            {
				HeapItem_t pa = this.Agents[i];
                foreach (Agent pA in pa.agents.Values)
                {
                    if (pA.IsActive())
                    {
                        pA.btexec();

                        //in case m_bTickAgents was set to false by pA's bt
                        if (!this.m_bTickAgents)
                        {
                            break;
                        }
                    }
                }
            }

            if (Agent.IdMask() != 0)
            {
                int contextId = this.GetContextId();
                Context c = Context.GetContext(contextId);

                c.LogStaticVariables(null);
            }
        }

        #region SocketLog
        public void LogCurrentStates()
        {
            string msg = string.Format("LogCurrentStates {0} {1}", this.GetClassTypeName(), this.Agents.Count);
            behaviac.Debug.Log(msg);

            //force to log vars
            foreach (HeapItem_t pa in this.Agents)
            {
                foreach (Agent pA in pa.agents.Values)
                {
                    if (pA.IsMasked())
                    {
                        pA.LogVariables(true);
                    }
                }
            }

            if (this.IsMasked())
            {
                this.LogVariables(true);
            }
        }
		#endregion
	}
}

File: src/world/world.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "behaviac/agent/context.h"
#include "behaviac/world/world.h"
#include "behaviac/base/logging/logging.h"
#include "behaviac/base/socket/socketconnect.h"
#include "behaviac/base/file/filesystem.h"
#include "behaviac/property/property.h"

#include "behaviac/base/core/profiler/profiler.h"

BEGIN_PROPERTIES_DESCRIPTION(behaviac::World)
	//
END_PROPERTIES_DESCRIPTION()

namespace behaviac
{
	World::World() : Agent(), m_bTickAgents(true)
	{
	}

	World::~World()
	{
		Context& c = Context::GetContext(this->m_context_id);
		World* pW = c.GetWorld(false);
		BEHAVIAC_UNUSED_VAR(pW);
		BEHAVIAC_ASSERT(pW == 0 || pW == this);
		c.SetWorld(0);
	}


	class BEHAVIAC_API DefaultWorld : public World
	{
	public:
		DECLARE_BEHAVIAC_OBJECT(DefaultWorld, World);

		DefaultWorld()
		{
		}

		virtual ~DefaultWorld()
		{
		}
	};


	BEGIN_PROPERTIES_DESCRIPTION(DefaultWorld)
		//
	END_PROPERTIES_DESCRIPTION()

		
	World* CreateDefaultWorld(int contextId)
	{
		World* pWorld = Agent::Create<DefaultWorld>(0, contextId);	

		return pWorld;
	}


	World* World::_GetWorld(int contextId)
	{
		Context& c = Context::GetContext(contextId);
		World* pWorld = c.GetWorld();

		BEHAVIAC_ASSERT(pWorld);

		return pWorld;
	}

	void World::AddAgent(Agent* pAgent)
	{
		ASSERT_MAIN_THREAD();

		int agentId = pAgent->GetId();
		int priority = pAgent->GetPriority();
		PriorityAgents_t::iterator it = std::find_if(this->m_agents.begin(), this->m_agents.end(), HeapFinder_t(priority));
		if (it == this->m_agents.end())
		{
			HeapItem_t pa;
			pa.priority = priority;
			pa.agents[agentId] = pAgent;
			this->m_agents.push_back(pa);
		}
		else
		{
			HeapItem_t& pa = *it;
			pa.agents[agentId] = pAgent;
		}
	}

	void World::RemoveAgent(Agent* pAgent)
	{
		ASSERT_MAIN_THREAD();

		int agentId = pAgent->GetId();
		int priority = pAgent->GetPriority();

		PriorityAgents_t::iterator it = std::find_if(this->m_agents.begin(), this->m_agents.end(), HeapFinder_t(priority));

		if (it != this->m_agents.end())
		{
			HeapItem_t& pa = *it;

			Agents_t::iterator ita = pa.agents.find(agentId);
			if (ita != pa.agents.end())
			{
				pa.agents.erase(ita);
			}
		}
	}

	void World::RemoveAllAgents()
	{
		ASSERT_MAIN_THREAD();

		this->m_agents.clear();
	}


	void World::LogCurrentStates()
	{
		if (Config::IsLoggingOrSocketing())
		{
			//force to log vars
			for (PriorityAgents_t::iterator it = this->m_agents.begin(); it != this->m_agents.end(); ++it)
			{
				HeapItem_t& pa = *it;

				for (Agents_t::iterator ita = pa.agents.begin(); ita != pa.agents.end(); ++ita)
				{
					Agent* pA = ita->second;

					if (this->IsMasked())
					{
						pA->LogVariables(true);
					}
				}
			}

			if (this->IsMasked())
			{
				this->LogVariables(true);
			}
		}
	}

	EBTStatus World::btexec()
	{
		behaviac::Profiler::GetInstance()->BeginFrame();

		Workspace::LogFrames();
		Workspace::HandleRequests();

		if (Workspace::GetAutoHotReload())
		{
			Workspace::HotReload();
		}

		EBTStatus s = super::btexec();
		BEHAVIAC_UNUSED_VAR(s);

		if (this->m_bTickAgents)
		{
			this->btexec_agents();
		}

		behaviac::Profiler::GetInstance()->EndFrame();

		//TODO: BT_INVALID
		return BT_RUNNING;
	}

	void World::ToggleTickAgents(bool bTickAgents)
	{
		this->m_bTickAgents = bTickAgents;
	}

	bool World::IsTickAgents() const
	{
		return this->m_bTickAgents;
	}

	void World::btexec_agents()
	{
		std::make_heap(this->m_agents.begin(), this->m_agents.end(), HeapCompare_t());

		for (PriorityAgents_t::iterator it = this->m_agents.begin(); it != this->m_agents.end(); ++it)
		{
			HeapItem_t& pa = *it;

			for (Agents_t::iterator ita = pa.agents.begin(); ita != pa.agents.end(); ++ita)
			{
				Agent* pA = ita->second;

				if (pA->IsActive())
				{
					pA->btexec();

					//in case m_bTickAgents was set to false by pA's bt
					if (!this->m_bTickAgents)
					{
						break;
					}
				}
			}
		}

		if (Agent::IdMask() != 0)
		{
			int contextId = this->GetContextId();
			Context& c = Context::GetContext(contextId);

			c.LogStaticVariables(0);
		}
	}

	void World::btreloadall()
	{
		Agent::btreloadall();

		for (PriorityAgents_t::iterator it = this->m_agents.begin(); it != this->m_agents.end(); ++it)
		{
			HeapItem_t& pa = *it;

			for (Agents_t::iterator ita = pa.agents.begin(); ita != pa.agents.end(); ++ita)
			{
				Agent* pA = ita->second;

				pA->btreloadall();
			}
		}	
	}

}//namespace behaviac
